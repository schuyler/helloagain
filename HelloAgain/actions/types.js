'use strict'

// Contacts
export const TOGGLE_ACTIVE = 'TOGGLE_ACTIVE'

// Friends
export const UPDATE_FRIEND = 'UPDATE_FRIEND'
