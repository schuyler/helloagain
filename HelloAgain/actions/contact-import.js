'use strict'

import Contacts from 'react-native-contacts'

import {
  CONTACTS_LOADED, CONTACT_LOAD_FAILED
} from './types';

export const contactsLoaded = (contacts) => {
  return {type: CONTACTS_LOADED, contacts: contacts}
}

export const contactLoadFailed = (error) => {
  return {type: CONTACT_LOAD_FAILED, error: error}
}

export const loadNativeContacts = () => {
  // Create a Promise that gets all native contacts and resolves on success.
  const loadContacts = new Promise((resolve, reject) => {
    Contacts.getAll((err, contacts) => {
      err ? reject(err) : resolve(contacts)
    })
  })
  // Return an action thunk that:
  //   1. Loads all contacts
  //   2. On success, dispatches the `contactsLoaded` action
  //   3. Then, dispatches the `writeIDsForAllFriends` to write back any new
  //      IDs to the native contact store
  //   4. On failure, dispatch the `contactLoadFailed` action
  return dispatch => {
    return loadContacts
      .then(contacts => {
        let result = dispatch(contactsLoaded(contacts))
        console.log("contactsLoaded:", Object.values(result).length)
      })
      .then(_ => { return dispatch(writeIDsForAllFriends()) })
      .catch(error => dispatch(contactLoadFailed(error)))
  }
}

const generateURL = (item) => {
  return "helloagain://to/" + item.helloAgainID
}

// Check a single contact record to see if it needs the HelloAgain URL
// written to it, and if it does, write it.
export const writeIDtoNativeContact = (item) => {
  // Never been active? Don't touch the native contact.
  if (item.isActive === undefined) {
    return
  }
  // Are any of the contact's urlAddresses a HelloAgain address? Great. Skip
  // this one.
  let matchingURLs = (item.urlAddresses || []).filter(url => url.label == "HelloAgain")
  if (matchingURLs.length > 0) {
    return
  }
  // Otherwise, generate the URL, push it onto the urlAddresses list with the
  // right label, and then update the native contact store, with a log message
  // either way.
  let url = generateURL(item)
  item.urlAddresses.push({label: "HelloAgain", urlAddress: url})
  Contacts.updateContact(item, (err, contact) => {
    if (err) {
      console.log(`Error updating ${item.givenName} ${item.familyName} ` +
                  `with URL ${url}:`, err, contact)
    } else {
      console.log(`Wrote URL to Contacts: ${url}`)
    }
  })
}

const writeIDsForAllFriends = () => {
  // Return an action thunk that is pure side-effects -- just iterate
  // over the entire set of contacts and make sure that any once-active friends
  // that don't have HelloAgain URLs associated with them get one. This
  // shouldn't really ever happen, because the `toggleActive` action takes care
  // of the records one-by-one, but I guess it can't hurt to be certain.
  return (_, getState) => {
    const state = getState()
    const friends = Object.values(state.friends).filter(f => f.isActive !== undefined)
    friends.forEach(item => writeIDtoNativeContact(item))
  }
}
