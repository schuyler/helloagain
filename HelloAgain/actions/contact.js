'use strict'

import { TOGGLE_ACTIVE } from './types';

export const toggleActive = (friend) => {
  return { type: TOGGLE_ACTIVE, friend: friend }
}
