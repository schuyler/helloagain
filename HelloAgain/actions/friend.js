'use strict'

import {
  UPDATE_FRIEND,
  MARK_AS_CONTACTED
} from './types'

export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}

export const markAsContacted = (friend) => {
  return { type: MARK_AS_CONTACTED, friend: friend }
}
