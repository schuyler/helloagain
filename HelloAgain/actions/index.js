'use strict';

// Enumerate the dispatchable actions.
export const UPDATE_FRIEND = 'UPDATE_FRIEND'

// Helpers for dispatching actions.
export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}
