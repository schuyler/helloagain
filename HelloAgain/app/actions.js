export const IMPORT_CONTACTS = 'IMPORT_CONTACTS';

export const UPDATE_FRIEND = 'UPDATE_FRIEND';

export const FRIEND_CONTACTED = 'FRIEND_CONTACTED';

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';

export function importContacts(contactList) {
  return { type: IMPORT_CONTACTS, contactList };
}

export function updateFriend(friend) {
  return { type: UPDATE_FRIEND, friend };
}
