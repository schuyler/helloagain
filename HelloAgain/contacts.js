'use strict';

import { Friends } from './models';
let AddressBook = require('react-native-contacts');

export function loadAllContacts() {
    // Load all the contacts from the system addressbook, and within that list,
    // substitute any contacts we already have in our database (and merging any
    // changes from the addressbook). We have to wait until the Contacts model
    // has loaded fully. Returns a Promise that is resolved when everything is
    // loaded.
    return new Promise((resolve, reject) => {
      Friends.hasLoaded.then(() => {
        AddressBook.getAll((err, contacts) => {
          if (err) {
            console.log("Reading addressbook failed:", err);
            reject(err);
            return;
          }
          // Update contacts from the addressbook with whatever we have stored in
          // our database.
          for (var i = 0; i < contacts.length; i++) {
            var storedContact = Friends.load(contacts[i].recordID);
            if (storedContact) {
              Object.assign(storedContact, contacts[i]);
              if (storedContact.isSelected) { // only write back changes to selected contacts
                Friends.save(storedContact);
              }
              contacts[i] = storedContact;
            }
          }
          resolve(contacts);
        });
      });
    });
};
