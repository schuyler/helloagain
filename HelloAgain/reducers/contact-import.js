'use strict'

import {
  CONTACTS_LOADED, CONTACT_LOAD_FAILED
} from '../actions/types';

const contactImport = (state = {}, action) => {
  switch (action.type) {
    case CONTACTS_LOADED:
      return {...state, count: action.contacts.length, loaded: true}
    case CONTACT_LOAD_FAILED:
      return {...state, error: action.error, loaded: false}
    default:
      return state
  }
}

export default contactImport
