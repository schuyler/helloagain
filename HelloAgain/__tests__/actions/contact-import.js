'use strict';

import configureMockStore from 'redux-mock-store'
import {CONTACT_FIXTURE} from '../fixtures/contacts'
import {CONTACTS_LOADED} from '../../actions/types'
import * as actions from '../../actions/contact-import'
import Contacts from 'react-native-contacts'
jest.mock('react-native-contacts')

const bob = {helloAgainID: "bob-dobbs-5555", givenName: "Bob", familyName: "Dobbs"}
const mockStore = configureMockStore()

describe('contactsLoaded', () => {
  it('should return an action', () => {
    expect(
      actions.contactsLoaded([bob])
    ).toMatchSnapshot()
  })
})

describe('contactLoadFailed', () => {
  it('should return an action', () => {
    expect(
      actions.contactLoadFailed("no idea why not")
    ).toMatchSnapshot()
  })
})

describe('loadNativeContacts', () => {
  const store = mockStore({})
  it('should dispatch CONTACTS_LOADED', () => {
    Contacts.getAll = jest.fn(_ => Promise.resolve(CONTACT_FIXTURE))
    actions.loadNativeContacts(store).then(() => {
        let dispatched = store.getActions()
        expect(dispatched[0].type).toBe(CONTACTS_LOADED)
        expect(dispatched[0].contacts.length).toBeGreaterThan(0)
      })
      .catch(err => console.error(err))
  })
})

describe('writeIDtoNativeContact', () => {
  it('should not write a never-active contact', () => {
    Contacts.updateContact = jest.fn()
    actions.writeIDtoNativeContact(bob)
    expect(Contacts.updateContact).not.toHaveBeenCalled()
  })
  it('should not write a contact that already has a URL', () => {
    Contacts.updateContact = jest.fn()
    actions.writeIDtoNativeContact({
      ...bob, 
      urlAddresses: [
        {label: "HelloAgain", urlAddress: "helloagain://to/bob-dobbs-5555"}
      ],
      isActive: true
    })
    expect(Contacts.updateContact).not.toHaveBeenCalled()
  })
  it('should write a contact that doesn\'t have a deep link', () => {
    for (let isActive of [true, false]) {
      let updatedContact = null
      Contacts.updateContact = jest.fn(contact => {updatedContact = contact})
      actions.writeIDtoNativeContact({...bob, urlAddresses: [], isActive: isActive})
      expect(Contacts.updateContact).toHaveBeenCalled()
      let url = updatedContact.urlAddresses[0]
      expect(url.label).toBe("HelloAgain")
      expect(url.urlAddress).toBe("helloagain://to/" + bob.helloAgainID)
    }
  })
})
