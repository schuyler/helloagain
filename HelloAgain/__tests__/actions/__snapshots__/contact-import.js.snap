exports[`contactLoadFailed should return an action 1`] = `
Object {
  "error": "no idea why not",
  "type": "CONTACT_LOAD_FAILED",
}
`;

exports[`contactsLoaded should return an action 1`] = `
Object {
  "contacts": Array [
    Object {
      "familyName": "Dobbs",
      "givenName": "Bob",
      "helloAgainID": "bob-dobbs-5555",
    },
  ],
  "type": "CONTACTS_LOADED",
}
`;
